server/mediaserver.go

```diff
@@ -130,10 +130,11 @@ func (s *LivepeerServer) StartMediaServer(ctx context.Context, transcodingOption
 	s.LPMS.HandleRTMPPlay(getRTMPStreamHandler(s))
 
 	//LPMS hanlder for handling HLS video play
-	s.LPMS.HandleHLSPlay(getHLSMasterPlaylistHandler(s), getHLSMediaPlaylistHandler(s), getHLSSegmentHandler(s))
+	s.LPMS.HandleHLSPlay(getHLSMasterPlaylistHandler(s), getHLSMediaPlaylistHandler(s), getHLSSegmentHandler(s), getSegmentPushHandler(s))
```

Comment:
I'd steer away from making changes to LPMS here.

LPMS can be parameterized with a custom HTTP mux, and this can be used to extend HTTP with additional endpoints. The orchestrator already does this:

https://github.com/livepeer/go-livepeer/blob/f5106de234ff99f26916f6965a0f8fe41e789f25/server/mediaserver.go#L103-L117

https://github.com/livepeer/go-livepeer/blob/f5106de234ff99f26916f6965a0f8fe41e789f25/server/rpc.go#L133-L145

Reply:
Would that mean including something like this in mediaserver.go, in under `StartMediaServer`? 
```golang
if s.LivepeerNode.NodeType == core.BroadcasterNode {
		s.HTTPMux.HandleFunc("/live/", func(w http.ResponseWriter, r *http.Request) {
			s.HandlePush(w, r)
		})
	}
```

If so, how would we then add the following, which currently live in player.go, to that same `HTTMux`, so all routes can be handled on that port? 
``` golang
        s.mux.HandleFunc("/stream/", func(w http.ResponseWriter, r *http.Request) {
		handleLive(w, r, getMasterPlaylist, getMediaPlaylist, getSegment)
	})

	s.mux.HandleFunc("/vod/", func(w http.ResponseWriter, r *http.Request) {
		handleVOD(r.URL, s.VodPath, w)
	})
```

`func (s *VidPlayer) HandleHLSPlay` in lpms/vidplayer/player.go does not have access to the Livepeer server, meaning no access to it's `HTTPMux` field.

Reply:
> Would that mean including something like this in mediaserver.go, in under StartMediaServer?

Yep, that looks about right. Things could likely be simplified to:

```golang
 opts := lpmscore.LPMSOpts{ 
 	RtmpAddr: rtmpAddr, RtmpDisabled: true, 
 	HttpAddr: httpAddr, 
 	WorkDir:  lpNode.WorkDir, 
 	HttpMux: http.NewServeMux(), // does it make sense to pass this in via NewLivepeerNode param?
 } 
switch lpNode.NodeType {
case core.BroadcasterNode:
  opts.RtmpDisabled = false;
  opts.HttpMux.HandleFunc("/live/", s.HandlePush)
}
```

> func (s *VidPlayer) HandleHLSPlay in lpms/vidplayer/player.go does not have access to the Livepeer server, meaning no access to it's HTTPMux field.

The LPMS vidplayer code remains untouched, since the `s.mux` would be the same as the `opts.HTTPMux` that's passed in to `lpmscore.New`. See the implementation:

https://github.com/livepeer/lpms/blob/4b2845787110924d95aff019d16174232b29d88c/core/lpms.go#L81

https://github.com/livepeer/lpms/blob/4b2845787110924d95aff019d16174232b29d88c/vidplayer/player.go#L43-L47

One day, I'd be happy to refactor LPMS to remove the vidplayer HTTP handling logic because it's barely used, but this will be fine for now.

Reply:
Hmmm but how would we have access to `s.HandlePush` in `NewLivepeerServer`, if we're creating the LivepeerServer at the end of that function? Even if we do create the LivepeerServer prior to that, we have to create the LPMS server prior to creating the LivepeerServer, which is the only time we can change/add to `lpmscore.LPMSOpts`, and therefore ... we don't have access to those properties from `go-livepeer`. 

We could do this: 
``` golang
func NewLivepeerServer(rtmpAddr string, httpAddr string, lpNode *core.LivepeerNode) *LivepeerServer {
	opts := lpmscore.LPMSOpts{
		RtmpAddr: rtmpAddr, RtmpDisabled: true,
		HttpAddr: httpAddr,
		WorkDir:  lpNode.WorkDir,
		HttpMux:  http.NewServeMux(),
	}
	switch lpNode.NodeType {
	case core.BroadcasterNode:
		opts.RtmpDisabled = false
	}
	server := lpmscore.New(&opts)
	s := LivepeerServer{RTMPSegmenter: server, LPMS: server, LivepeerNode: lpNode, HTTPMux: opts.HttpMux, connectionLock: &sync.RWMutex{}, rtmpConnections: make(map[core.ManifestID]*rtmpConnection)}
	s.HTTPMux.HandleFunc("/live/", func(w http.ResponseWriter, r *http.Request) { //out
		s.HandlePush(w, r)
	})
	return &s
}
```

However, that doesn't work in practice. 8935 still does not serve `/live` in that scenario. Adding a route to  `s.HTTPMux` does not seem to carry over to `opts.HttpMux`. I'm not sure I understand how we can avoid changing LPMS code to make this work :(

Reply:
> We could do this:

Yes, that way will work. I'd still avoid wrapping it in an intermediate trampoline-like function and do it like this:

```
s.HTTPMux.HandleFunc("/live/", s.HandlePush)
```

> 8935 still does not serve /live in that scenario.

Does it serve `/stream/` at all? If you look at the [LPMS implementation](https://github.com/livepeer/lpms/blob/master/core/lpms.go#L96-L100) [1],  LPMS also starts the HTTP server if a default mux is used. If a mux is passed in, then the HTTP server attached to that needs to be started by someone else. So I suspect right now there isn't even a running server on 8935.

In that case, the server will need to be started somewhere on the broadcaster side, eg `s.HttpMux.ListenAndServe`

Reply:
@angyangie it should look like this: 41535c6947a22d5349ceb92e94f27e5e08f9162e

Reply:
THANK YOU @darkdarkdragon and @j0sh! No server was running on 8935, so your solution works well @darkdarkdragon